File: qubgrademe-marksfornext/tests/unittests/validate.test.js
const { validate, isStringPositiveInteger } = require("../../src/validate");
const { assert, expect } = require("chai");
const should = require("chai").should();

describe("Testing integer conversion", () => {
    test("'100' is a valid integer", () => {
        const result = isStringPositiveInteger("100");
        expect(result).to.be.true;
    });

    test("'0' is a valid integer", () => {
        const result = isStringPositiveInteger("0");
        expect(result).to.be.true;
    });

    test("'50' is a valid integer", () => {
        const result = isStringPositiveInteger("50");
        expect(result).to.be.true;
    });

    test("'101' is not valid integer", () => {
        const result = isStringPositiveInteger("101");
        expect(result).to.be.false;
    });

    test("'-1' is not valid integer", () => {
        const result = isStringPositiveInteger("-1");
        expect(result).to.be.false;
    });

    test("undefined is not valid integer", () => {
        const result = isStringPositiveInteger(undefined);
        expect(result).to.be.false;
    });

    test("null is not valid integer", () => {
        const result = isStringPositiveInteger(null);
        expect(result).to.be.false;
    });

    test("NaN is not valid integer", () => {
        const result = isStringPositiveInteger(NaN);
        expect(result).to.be.false;
    });
});

describe("Validates the query params to ensure a valid request has been sent", () => {
    test("Succeeds with full array of modules and marks", () => {
        const modules = [
            "CSC 3065",
            "CSC3021",
            "CSC3063",
            "CSC3059",
            "CSC3069",
        ];
        const marks = ["80", "70", "70", "80", "60"];

        const result = validate(modules, marks);
        result.should.have.property("success").equals(true);
        result.should.have.property("message").equals("");
    });

    test("Succeeds with modules and marks missing at the end", () => {
        const modules = [
            "CSC 3065",
            "CSC3021",
            "CSC3063",
            "CSC3059",
            undefined,
        ];

        const marks = ["80", "70", "70", "80", undefined];

        const result = validate(modules, marks);
        result.should.have.property("success").equals(true);
        result.should.have.property("message").equals("");
    });

    test("Succeeds with modules and marks missing at the start", () => {
        const modules = [
            undefined,
            "CSC3021",
            "CSC3063",
            "CSC3059",
            "CSC 3065",
        ];

        const marks = [undefined, "70", "70", "80", "90"];

        const result = validate(modules, marks);
        result.should.have.property("success").equals(true);
        result.should.have.property("message").equals("");
    });

    test("Fails when a module has a mark of over 100", () => {
        const modules = [
            "CSC 3068",
            "CSC3021",
            "CSC3063",
            "CSC3059",
            "CSC 3065",
        ];

        const marks = ["30", "70", "70", "80", "101"];

        const result = validate(modules, marks);
        result.should.have.property("success").equals(false);
        result.should.have
            .property("message")
            .equals("Please provide a valid integer for every entered module");
    });

    test("Fails when a module has a mark below 0", () => {
        const modules = [
            "CSC 3068",
            "CSC3021",
            "CSC3063",
            "CSC3059",
            "CSC 3065",
        ];

        const marks = ["30", "70", "70", "80", "-3"];

        const result = validate(modules, marks);
        result.should.have.property("success").equals(false);
        result.should.have
            .property("message")
            .equals("Please provide a valid integer for every entered module");
    });

    test("Fails when a mark is provided without a module name", () => {
        const modules = [
            undefined,
            "CSC3021",
            "CSC3063",
            "CSC3059",
            "CSC 3065",
        ];

        const marks = ["30", "70", "70", "80", "10"];

        const result = validate(modules, marks);
        result.should.have.property("success").equals(false);
        result.should.have
            .property("message")
            .equals("Please provide a module name for all marks entered");
    });

    test("Fails when all modules are empty", () => {
        const modules = [undefined, undefined, undefined, undefined, undefined];

        const marks = ["30", "70", "70", "80", "10"];

        const result = validate(modules, marks);
        result.should.have.property("success").equals(false);
        result.should.have
            .property("message")
            .equals(
                "Please provide some module codes and their respective marks",
            );
    });
});

File: qubgrademe-classifygrade-python/tests/validate_test.py
from src.validation import validate, is_string_positive_integer

# class Test_is_string_positive_integer():
#     def test_10_string_is_valid(self):
#         result = is_string_positive_integer("10")
#         assert result == True

#     def test_10_int_is_valid(self):
#         result = is_string_positive_integer(10)
#         assert result == True

#     def none_is_not_valid(self):
#         result = is_string_positive_integer(10)
#         assert result == True

#     def test_string_above_100_is_not_valid(self):
#         result = is_string_positive_integer("101")
#         assert result == False

#     def test_number_above_100_is_not_valid(self):
#         result = is_string_positive_integer(101)
#         assert result == False

#     def test_string_below_0_is_not_valid(self):
#         result = is_string_positive_integer("-3")
#         assert result == False

#     def test_number_below_0_is_not_valid(self):
#         result = is_string_positive_integer(-3)
#         assert result == False


class Test_validate():

    # def test_succeeds_with_all_modules_and_valid_marks(self):
    #     modules = [
    #             "CSC 3065",
    #             "CSC3021",
    #             "CSC3063",
    #             "CSC3059",
    #             "CSC3069",
    #         ]
    #     marks = ["80", "70", "70", "80", "60"]

    #     success, message = validate(modules, marks)
    #     assert success == True
    #     assert message == ""

    # def test_succeeds_with_missing_pair_of_modules_and_marks(self):
    #     modules = [
    #             "CSC 3065",
    #             "CSC3021",
    #             "CSC3063",
    #             "CSC3059",
    #             None,
    #         ]
    #     marks = ["80", "70", "70", "80", None]

    #     success, message = validate(modules, marks)
    #     assert success == True
    #     assert message == ""

    #     modules = [
    #             "CSC 3065",
    #             "CSC3021",
    #             "CSC3063",
    #             "CSC3059",
    #             "CSC3069",
    #         ]
    #     marks = ["80", "70", "70", "80", "60"]

    #     success, message = validate(modules, marks)
    #     assert success == True
    #     assert message == ""

    # def test_succeeds_with_missing_pair_of_modules_and_marks2(self):
        # modules = [
        #         None,
        #         "CSC3021",
        #         "CSC3063",
        #         "CSC3059",
        #         None,
        #     ]
        # marks = ["80", "70", "70", "80", None]

        # success, message = validate(modules, marks)
        # assert success == True
        # assert message == ""

        # modules = [
        #         None,
        #         "CSC3021",
        #         "CSC3063",
        #         "CSC3059",
        #         "CSC3069",
        #     ]
        # marks = ["80", "70", "70", "80", "60"]

        # success, message = validate(modules, marks)
        # assert success == True
        # assert message == ""

    def test_fails_with_mark_not_present_for_module(self):
        modules = [
                "CSC 3065",
                "CSC 3021",
                "CSC 3063",
                "CSC 3059",
                "CSC 3069",
            ]
        marks = ["80", "70", "70", "80", None]

        success, message = validate(modules, marks)
        assert success == False
        assert message == "Please provide a valid integer for every entered module" 
    
    def test_fails_with_mark_over_100(self):
        modules = [
                "CSC 3065",
                "CSC 3021",
                "CSC 3063",
                "CSC 3059",
                "CSC 3069",
            ]
        marks = ["80", "70", "70", "80", "120"]

        success, message = validate(modules, marks)
        assert success == False
        assert message == "Please provide a valid integer for every entered module" 

    def test_fails_with_mark_below_0(self):
        modules = [
                "CSC 3065",
                "CSC 3021",
                "CSC 3063",
                "CSC 3059",
                "CSC 3069",
            ]
        marks = ["80", "70", "70", "80", "-1"]

        success, message = validate(modules, marks)
        assert success == False
        assert message == "Please provide a valid integer for every entered module" 

    def test_fails_with_module_not_present_for_mark(self):
        modules = [
                "CSC 3065",
                "CSC 3021",
                "CSC 3063",
                "CSC 3059",
                None,
            ]
        marks = ["80", "70", "70", "80", "30"]

        success, message = validate(modules, marks)
        assert success == False
        assert message == "Please provide a module name for all marks entered" 

    def test_fails_if_no_modules_are_entered(self):
        modules = [
                None,
                None,
                None,
                None,
                None,
            ]
        marks = ["80", "70", "70", "80", "30"]

        success, message = validate(modules, marks)
        assert success == False
        assert message == "Please provide some module codes and their respective marks" 

    def test_fails_if_no_modules_are_entered_empty_string(self):
        modules = [
                "",
                "",
                "",
                "",
                "",
            ]
        marks = ["80", "70", "70", "80", "30"]

        success, message = validate(modules, marks)
        assert success == False
        assert message == "Please provide some module codes and their respective marks" 
    
    def test_fails_if_no_modules_are_entered_whitespace(self):
        modules = [
                " ",
                " ",
                " ",
                " ",
                " ",
            ]
        marks = ["80", "70", "70", "80", "30"]

        success, message = validate(modules, marks)
        assert success == False
        assert message == "Please provide some module codes and their respective marks" 

    def test_fails_mix_cases(self):
        modules = [
                "CSC 3065",
                "CSC 3021",
                "CSC 3063",
                " ",
                "CSC 3069",
            ]
        marks = ["80", "70", "70", "80", None]

        success, message = validate(modules, marks)
        assert success == False
        assert message == "Please provide a module name for all marks entered" 


    def test_fails_mix_cases2(self):
        modules = [
                "CSC 3065",
                "CSC 3021",
                "CSC 3063",
                "CSC 3068",
                " ",
            ]
        marks = ["80", "70", "70", None, "80"]

        success, message = validate(modules, marks)
        assert success == False
        assert message == "Please provide a valid integer for every entered module" 